### ISSUE TITLE
Export "npm run rarity"

### ISSUE THREAD

**Issue by joaogfmoreira** [HashLips/hashlips_art_engine#1025]
Hello

I generated a 10k collection and am trying to export the "run rarity" from the terminal (VS code). but it won't display all the traits in the terminal. how can i see the full output?

**Comment by joaogfmoreira** [comment]
managed to get the full output by running it directly on the Macbook Terminal app. would be nice to know how to get it on VS Code though

**Comment by jimsabo** [comment]
It would be great if the original individual JSON file could be updated after the run to include the rarity data, or dump the rarity info to a JSON file at the end of the run. My end goal is to show how rare or not rare an individual file wound up being; anything that gets me there faster is great.

**Comment by bolshoytoster** [comment]
@jimsabo `node utils/rarity.js` already gives rarity, here:
https://github.com/HashLips/hashlips_art_engine/blob/d8ee279043d2d4a8de3bdfac0d89d0e966fb04a2/utils/rarity.js#L75-L82

You could format it as a json object and write it to a file.

**Comment by joaogfmoreira** [comment]
> @jimsabo `node utils/rarity.js` already gives rarity, here:
> 
> https://github.com/HashLips/hashlips_art_engine/blob/d8ee279043d2d4a8de3bdfac0d89d0e966fb04a2/utils/rarity.js#L75-L82
> 
> You could format it as a json object and write it to a file.

Thanks. It indeed prints it to the console, but in the case of VS Code, it won't let us see the full output. I managed to see in the terminal. How could I format it as a json and write it to a file?

**Comment by bolshoytoster** [comment]
@joaogfmoreira when I get home, I'll have a go.

**Comment by bolshoytoster** [comment]
@joaogfmoreira you can just add
```js
fs.writeFileSync(
    `${basePath}/foo.json`,
    rarityData,
    () => {},
)
```
To the end of `utils/rarity.js`.

**Comment by Pwnkmrdroch** [comment]
How can I get Rarity of complete collection incase I created multiple cases (10+).
Right now I am using node utils/rarity.js but it is showing a rarity of first case only.
Can anyone help asap?

Thank you

**Comment by buhduhzah** [comment]
`npm run rarity > rarity.log 2>&1` has worked for me. it creates a txt file

**Comment by Pwnkmrdroch** [comment]
Thank you @buhduhzah 
Can you let me know what are these 2>&1 values as I have 10 cases for "growEditionSizeTo" so for me it is not exporting all, just for initial case.

**Comment by Pwnkmrdroch** [comment]
![Screenshot 2022-05-10 193821](https://user-images.githubusercontent.com/102255858/167648502-7859a598-5059-40d2-a194-6df884ff807e.png)
here is a screenshot attacted.

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch
> Can you let me know what are these 2>&1 values

`2>&1` means to send `stderr` to `stdout`.
It's irrelevant here and I don't think powershell supports this and you can ignore it.

> for me it is not exporting all

Do you mean you only have the first configuration in your collection or only the first one shows up on rarity?

The issue may be that you use `displayName`, which means the rarity script can't know what layer it actual is.

**Comment by Pwnkmrdroch** [comment]
Yes, only the first one shows up on rarity.

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch have you tried not using `displayName`? It means that the rarity script can't know what layer it actual is.

**Comment by Pwnkmrdroch** [comment]
No, actually i generate and delivered this project already so they are asking for this stuff now.

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch sorry, the rarity script doesn't currently support this. I can work on a version that doesn't need `layersConfiguration` though.

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch here it is:
```js
const basePath = process.cwd();
const fs = require("fs");

// read json data
let rawdata = fs.readFileSync(`${basePath}/build/json/_metadata.json`);
let data = JSON.parse(rawdata);
let editionSize = data.length;

let rarityData = [];

// fill up rarity chart with occurrences from metadata
data.forEach((element) => {
  let attributes = element.attributes;
  attributes.forEach((attribute) => {
    let traitType = attribute.trait_type;
    let value = attribute.value;
    let found = false;

    if (rarityData[traitType]) {
      let rarityDataTraits = rarityData[traitType];
      rarityDataTraits.forEach((rarityDataTrait) => {
        if (rarityDataTrait.trait == value) {
          // keep track of occurrences
          rarityDataTrait.occurrence++;
          found = true;
        }
      });
    } else {
      rarityData[traitType] = [];
    }

    if (!found) {
      // just get name for each element
      let rarityDataElement = {
        trait: value,
        occurrence: 1, // initialize at 1
      };
      rarityData[traitType].push(rarityDataElement);
    }
  });
});

// convert occurrences to occurence string
for (var layer in rarityData) {
  for (var attribute in rarityData[layer]) {
    // get chance
    let chance =
      ((rarityData[layer][attribute].occurrence / editionSize) * 100).toFixed(2);

    // show two decimal places in percent
    rarityData[layer][attribute].occurrence =
      `${rarityData[layer][attribute].occurrence} in ${editionSize} editions (${chance} %)`;
  }
}

// print out rarity data
for (var layer in rarityData) {
  console.log(`Trait type: ${layer}`);
  for (var trait in rarityData[layer]) {
    console.log(rarityData[layer][trait]);
  }
  console.log();
}
```

**Comment by Pwnkmrdroch** [comment]
wohooo, Thank you so much.... looks like working...
Is there any way club it with current metadata or I need to deliver it as sperate to developer?

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch
> Is there any way club it with current metadata

What fo you mean by 'club'?

**Comment by Pwnkmrdroch** [comment]
![PNG](https://user-images.githubusercontent.com/102255858/168264067-4b3f918c-12da-4bdb-bd83-e72ba8511c7e.png)
Sorry @buhduhzah , but this is still not exporting right details, missing some of traits like i have 2 folder of 16 clothes and it is exporting only 10 in rarity export. Same issue is with some other folders as well.

Do I have to changes in setting from code you shared above?

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch Make sure you've definetely using the `rarity.js` I posted. Could you paste the output of running it here please.

**Comment by Pwnkmrdroch** [comment]
@buhduhzah thank you for quick response. Yes, it was my bad, I did create a new copy but name was rarity_new.js which was why it not calculate right, but as I delete previous rarity file and added your new, it works perfect.

Thank you so much for your help.
Please, send me your wallet, I will send token of thanks.

**Comment by bolshoytoster** [comment]
@Pwnkmrdroch you're welcome.

BTC: 3PcvzFiZVgDSaEjVhBufc9qTbSgpNhWQAA

**Comment by Omegacris** [comment]
Hi all,
 
Similar to the cmd `npm run rarity > rarity.log 2>&1` what is the cmd to export into a file what are the Top 10 most rarest generated items (images) from the entire generated collection ?
Basically I'd like to have generated a Top 10 or 100 List showing the **Name** of the generated image and what **Traits** it contains (Background, Eyes, Mouth) types. If not Top # then the entire generated collection (e.g. 5000) but sorted from top to bottom with Most rare to Least rare.

Is it possible with a command line or maybe adding a procedure in the `utils/rarity.js` ?

**Comment by warengonzaga** [comment]
> @Pwnkmrdroch here it is:
> 
> ```js
> const basePath = process.cwd();
> const fs = require("fs");
> 
> // read json data
> let rawdata = fs.readFileSync(`${basePath}/build/json/_metadata.json`);
> let data = JSON.parse(rawdata);
> let editionSize = data.length;
> 
> let rarityData = [];
> 
> // fill up rarity chart with occurrences from metadata
> data.forEach((element) => {
>   let attributes = element.attributes;
>   attributes.forEach((attribute) => {
>     let traitType = attribute.trait_type;
>     let value = attribute.value;
>     let found = false;
> 
>     if (rarityData[traitType]) {
>       let rarityDataTraits = rarityData[traitType];
>       rarityDataTraits.forEach((rarityDataTrait) => {
>         if (rarityDataTrait.trait == value) {
>           // keep track of occurrences
>           rarityDataTrait.occurrence++;
>           found = true;
>         }
>       });
>     } else {
>       rarityData[traitType] = [];
>     }
> 
>     if (!found) {
>       // just get name for each element
>       let rarityDataElement = {
>         trait: value,
>         occurrence: 1, // initialize at 1
>       };
>       rarityData[traitType].push(rarityDataElement);
>     }
>   });
> });
> 
> // convert occurrences to occurence string
> for (var layer in rarityData) {
>   for (var attribute in rarityData[layer]) {
>     // get chance
>     let chance =
>       ((rarityData[layer][attribute].occurrence / editionSize) * 100).toFixed(2);
> 
>     // show two decimal places in percent
>     rarityData[layer][attribute].occurrence =
>       `${rarityData[layer][attribute].occurrence} in ${editionSize} editions (${chance} %)`;
>   }
> }
> 
> // print out rarity data
> for (var layer in rarityData) {
>   console.log(`Trait type: ${layer}`);
>   for (var trait in rarityData[layer]) {
>     console.log(rarityData[layer][trait]);
>   }
>   console.log();
> }
> ```

Added this to my own version of art engine. https://github.com/warengonzaga/thirdweb-art-engine/issues/6
